src/riot-rs-threads/src/sync/mutex.rs

```diff
@@ -0,0 +1,193 @@
+    fn locked_with_current(cs: CriticalSection) -> Self {
+        let (owner_id, owner_prio) = THREADS.with_mut_cs(cs, |mut threads| {
+            let current = threads
+                .current()
+                .expect("Function should be called inside a thread context.");
+            (current.pid, current.prio)
+        });
+        LockState::Locked {
+            waiters: ThreadList::new(),
+            owner_id,
+            owner_prio,
+        }
+    }
+}
+
+impl<T> Mutex<T> {
+    /// Creates a new **unlocked** [`Mutex`].
+    pub const fn new(value: T) -> Self {
+        Self {
+            state: UnsafeCell::new(LockState::Unlocked),
+            inner: UnsafeCell::new(value),
+        }
+    }
+
+    /// Returns whether the mutex is locked.
+    pub fn is_locked(&self) -> bool {
+        critical_section::with(|_| {
+            let state = unsafe { &*self.state.get() };
+            !matches!(state, LockState::Unlocked)
+        })
+    }
+
+    /// Acquires a mutex, blocking the current thread until it is able to do so.
+    ///
+    /// If the mutex was unlocked, it will be locked and a [`MutexGuard`] is returned.
+    /// If the mutex is locked, this function will block the current thread until the mutex gets
+    /// unlocked elsewhere.
+    ///
+    /// If the current owner of the mutex has a lower priority than the current thread, it will inherit
+    /// the waiting thread's priority.
+    /// The priority is reset once the mutex is released. This means that a **user can not change a thread's
+    /// priority while it holds the lock**, because it will be changed back after release!
+    ///
+    /// # Panics
+    ///
+    /// Panics if called outside of a thread context.
+    pub fn lock(&self) -> MutexGuard<T> {
+        critical_section::with(|cs| {
+            let state = unsafe { &mut *self.state.get() };
+            match state {
+                LockState::Unlocked => {
+                    *state = LockState::locked_with_current(cs);
+                }
+                LockState::Locked {
+                    waiters,
+                    owner_id,
+                    owner_prio,
+                } => {
+                    // Insert thread in waitlist, which also triggers the scheduler.
+                    match waiters.put_current(cs, ThreadState::LockBlocked) {
+                        // `Some` when the inserted thread is the highest priority
+                        // thread in the waitlist.
+                        Some(waiter_prio) if waiter_prio > *owner_prio => {
+                            // Current mutex owner inherits the priority.
+                            THREADS.with_mut_cs(cs, |mut threads| {
+                                threads.set_priority(*owner_id, waiter_prio)
+                            });
+                        }
+                        _ => {}
+                    }
+                    // Context switch happens here as soon as we leave the critical section.
```

Comment:
I put the comment down here on purpose because when first reading the code, it took me a while to understand that after the `put_current` call a context switch happens, and that's why in the code that then follows it can be assumed that the lock is now acquired (because the context switch back to the thread will only then happen).